### PR TITLE
fix: several bugs on games/tic_tac_toe.c

### PR DESCRIPTION
**Pull request by kana800**
#### Description of Change

Issue: #729 
As **mentioned** in the issue , The piece code listed below throws the program into an loop (from `line 247`) if an **invalid** character is entered.
```diff
        else
        {
+            int n = check_placex();
-             printf("Invalid move\n");

-            printf("Enter new position : ");
-            scanf("%d", &n1);
+           placex(n);
-            placex(n1);
        }
    }
    else
    {
-        printf("Invalid move \n");
+       int n = check_placex();
-        printf("Enter new position : ");
-        scanf("%d", &n1);
+       placex(n);
-        placex(n1);
    }
}
```
##### Added an array [dynamic] at `line36` :
```diff
+ int *stored_pos = NULL;
```
`line 47`:
```diff
+  stored_pos = (int*)calloc(9,sizeof(int))
```
##### _Assignment_ of the **stored_pos** array:

`line 331`, `line 308` & `line 278`  

example: 

```diff
+ stored_pos[e] = 1;
```
##### Freed the _stored_pos_ memory

`line 75`
```diff
+ free(stored_pos);
```

##### Added a new function called `check_places` at `line237` to `line256`

```C
int check_placex(){
	char input[50];
	int n1;
	while (1){
		fgets(input,49,stdin);
		if ( strlen(input) > 2 || strlen(input)  == 0){
			fprintf(stderr,"Invalid move, Enter number 1 - 9: ");
			continue;
		}
		if(sscanf(input,"%d",&n1) != 1){
			fprintf(stderr,"Invalid move, Enter number 1 - 9: ");
			continue;
		} 
		if ((stored_pos[n1-1] == 1) || (n1== 0)){
			fprintf(stderr,"Already allocated, Enter number: ");
			continue;
		}
		return n1;
	}
}	
```

The `while-loop` in the above above function will ask for an input until it gets a correct output.

1. The first **if condition** checks whether the **input** is within the range. _(strlen() is used so `string.h` lib is need)_
1. Second **if condition** checks if the value is integer
1. Third **if condition** checks if the value (_posistion_) is already occupied.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C/pull/771"></a>

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C/pull/771"></a>

### CODE REVIEW

**Review by Panquesito7 — APPROVED** [review]
LGTM. Thank you for your contribution! 🎉 👍

File: games/tic_tac_toe.c [hunk 1]
```diff
@@ -28,6 +29,12 @@ int checkwin();  // checks everytime when a player or computer places 'X' or 'O'
  */
 static char game_table[9];
 
+/**
+ * stored_pos array will store posistion of that were already allocated by the user
+ * calloc(9,sizeof(int)) means an array of 9 elements.
+ */
+int *stored_pos = NULL;
```
Comment (kvedala) [inline games/tic_tac_toe.c:36]:
```suggestion
static unsigned char *stored_pos = NULL;
```
* `static` to recognize the variable as local to this file. 
* `unsigned char` pointer to be same as line 30
Is this even required? isnt it serving the same purpose as `game_table`?

**Review by kvedala — CHANGES_REQUESTED** [review]
a simple fix for the bug would be to simply ensure that the input number is >=0 & <=9. There is no need to make so many changes.

**Comment by kana800** [comment]
> a simple fix for the bug would be to simply ensure that the input number is >=0 & <=9. There is no need to make so many changes.

I went a bit overboard with it

File: games/tic_tac_toe.c (hunk 1 above)
Comment (kana800) [reply games/tic_tac_toe.c:36]: I can't actually remember what I have done here, but I think the purpose of `stored_pos` was to make sure that the values that was inserted by the user was not overwritten.

**Comment by kana800** [comment]
@Panquesito7 any idea why the autochecks are failing ? 
I read the logs and I don't get it

**Comment by kvedala** [comment]
> > a simple fix for the bug would be to simply ensure that the input number is >=0 & <=9. There is no need to make so many changes.
> 
> I went a bit overboard with it
Good, it would be better to add your new contribution as a "v2" of the program, a new file as described in the contribution guidelines while keeping the fix to this code to its minimal.

**Comment by kana800** [comment]
> > > a simple fix for the bug would be to simply ensure that the input number is >=0 & <=9. There is no need to make so many changes.
> > 
> > 
> > I went a bit overboard with it
> 
> Good, it would be better to add your new contribution as a "v2" of the program, a new file as described in the contribution guidelines while keeping the fix to this code to its minimal.

Kept everything to minimal, removed `stored_pos` and used `game_table` . Instead of using `scanf` i went with another approach (line : `230`), `scanf` will take the `\n` of the input buffer.

File: games/tic_tac_toe.c [hunk 2]
```diff
@@ -218,6 +220,38 @@ void doublemode()
     }
 }
 
+/**
+ * @brief Helper function of place_x,place_y function
+ *
+ * @param None
```
Comment (Panquesito7) [inline games/tic_tac_toe.c:226]: No need to add this if there are no parameters.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C/runs/1505777581) warnings.

File: games/tic_tac_toe.c (hunk 2 above)
Comment (kana800) [reply games/tic_tac_toe.c:226]: okay

**Comment by kana800** [comment]
@Panquesito7 please check the `windows log`

File: games/tic_tac_toe.c [hunk 3]
```diff
@@ -218,6 +220,31 @@ void doublemode()
...
+	}
+}	
+
+
+
+
+
 /**
  * @brief Update table by placing an `X`
  *
```
Comment (Panquesito7) [inline games/tic_tac_toe.c:250]: In line 257, replace this line with this one:

```diff
- int n1;
+ int n1 = 0;
```

File: games/tic_tac_toe.c [hunk 4]
```diff
@@ -236,22 +263,14 @@ void placex(int m)
...
-
-        printf("Enter new position : ");
-        scanf("%d", &n1);
-
-        placex(n1);
+		int n = check_placex();
+		placex(n);
     }
 }
 /**
```
Comment (Panquesito7) [inline games/tic_tac_toe.c:276]: In line 308, replace this line with the following line:

```diff
- int n1;
+ int n1 = 0;
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please see https://github.com/TheAlgorithms/C/runs/1581156179#step:4:1407.

**Comment by kana800** [comment]
> Please see https://github.com/TheAlgorithms/C/runs/1581156179#step:4:1407.

Thanks a lot !

**Comment by Panquesito7** [comment]
https://github.com/TheAlgorithms/C/runs/1583568825#step:4:1424

**Comment by kana800** [comment]
```
2020-12-22T16:06:10.6979145Z   tic_tac_toe.c
2020-12-22T16:06:10.7562749Z C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\ucrt\stdio.h(378,9): warning C5105: macro expansion producing 'defined' has undefined behavior [D:\a\C\C\build\games\tic_tac_toe.vcxproj]
2020-12-22T16:06:10.7689739Z C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\ucrt\stdio.h(2437,9): warning C5105: macro expansion producing 'defined' has undefined behavior [D:\a\C\C\build\games\tic_tac_toe.vcxproj]
2020-12-22T16:06:10.7722359Z C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\ucrt\corecrt_search.h(188,5): warning C5105: macro expansion producing 'defined' has undefined behavior [D:\a\C\C\build\games\tic_tac_toe.vcxproj]
2020-12-22T16:06:10.7778834Z C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\ucrt\stdlib.h(79,5): warning C5105: macro expansion producing 'defined' has undefined behavior [D:\a\C\C\build\games\tic_tac_toe.vcxproj]
2020-12-22T16:06:10.7856859Z C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\ucrt\stdlib.h(1286,5): warning C5105: macro expansion producing 'defined' has undefined behavior [D:\a\C\C\build\games\tic_tac_toe.vcxproj]
2020-12-22T16:06:10.7904488Z C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\ucrt\time.h(589,5): warning C5105: macro expansion producing 'defined' has undefined behavior [D:\a\C\C\build\games\tic_tac_toe.vcxproj]
2020-12-22T16:06:10.7935445Z C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\ucrt\corecrt_memory.h(76,5): warning C5105: macro expansion producing 'defined' has undefined behavior [D:\a\C\C\build\games\tic_tac_toe.vcxproj]
2020-12-22T16:06:10.7984994Z C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\ucrt\corecrt_wstring.h(573,5): warning C5105: macro expansion producing 'defined' has undefined behavior [D:\a\C\C\build\games\tic_tac_toe.vcxproj]
2020-12-22T16:06:10.8038939Z C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\ucrt\string.h(531,5): warning C5105: macro expansion producing 'defined' has undefined behavior [D:\a\C\C\build\games\tic_tac_toe.vcxproj]

```
@Panquesito7 
Fixed Errors from my end, anything wrong with windows version of **LGTM** analysis?

**Comment by Panquesito7** [comment]
> anything wrong with windows version of **LGTM** analysis?

LGTM seems to be working fine (no errors/bugs reported).
Though I still see some Windows [warnings](https://github.com/TheAlgorithms/C/pull/771/checks?check_run_id=1595787547#step:4:1417) reported by the CI.

**Comment by Panquesito7** [comment]
My bad! It seems it has something to do with latest commit in the repository (https://github.com/TheAlgorithms/C/commit/b2def5ca0e6d8b97762d549516234fdd80568750).
Your code seems to be good, though. I do not see any errors reported. Only warnings, though.

**Comment by kvedala** [comment]
> My bad! It seems it has something to do with latest commit in the repository ([b2def5c](https://github.com/TheAlgorithms/C/commit/b2def5ca0e6d8b97762d549516234fdd80568750)).
> Your code seems to be good, though. I do not see any errors reported. Only warnings, though.

The CI fix is present here: #796 
Once this is merged,
The pull-requests and/or commits since Nov 29, 2020 can be rechecked by force triggering the CI checks on them, including this one

**Comment by kana800** [comment]
Recheck done 👍🏽 , Everything seems to be working alright

File: games/tic_tac_toe.c [hunk 5]
```diff
@@ -34,7 +35,8 @@ static char game_table[9];
  * @note No checks are included for program execution failures!
  */
 int main()
-{   srand(time(NULL));
+{   
+    srand( (unsigned int)time(NULL));
     int l = 0;
     do
```
Comment (north8side) [inline games/tic_tac_toe.c:41]:
```
,)•(

```

File: games/tic_tac_toe.c [hunk 6]
```diff
@@ -34,7 +35,8 @@ static char game_table[9];
  * @note No checks are included for program execution failures!
  */
 int main()
-{   srand(time(NULL));
+{   
+    srand( (unsigned int)time(NULL));
```
Comment (north8side) [inline games/tic_tac_toe.c:39]: ,(•):_

File: games/tic_tac_toe.c [hunk 7]
```diff
@@ -218,6 +220,31 @@ void doublemode()
...
+int check_placex(){
+	char input[50];
+	int n1;
+	while (1){
+		fgets(input,49,stdin);
+		if ( strlen(input) > 2 || strlen(input)  == 0){
+			fprintf(stderr,"Invalid move, Enter number 1 - 9: ");
+			continue;
+		}
+		if(sscanf(input,"%d",&n1) != 1){
```
Comment (north8side) [inline games/tic_tac_toe.c:232]: }{

**Pull request merged**